### PR TITLE
Support metastore copy between two different hive metastores

### PR DESCRIPTION
**Pull request by guy-har**
Part of #1674

### CODE REVIEW

File: cmd/lakectl/cmd/metastore.go [hunk 1]
```diff
@@ -208,6 +248,17 @@ func init() {
...
 	_ = metastoreDiffCmd.Flags().String("to-table", "", "destination table name [default is from-table]")
 
+	metastoreCmd.AddCommand(metastoreCopyAllCmd)
+
+	_ = metastoreCopyAllCmd.Flags().String("from-address", "", "source metastore address")
+	_ = metastoreCopyAllCmd.MarkFlagRequired("from-address")
+	_ = metastoreCopyAllCmd.Flags().String("to-address", "", "destination metastore address")
+	_ = metastoreCopyAllCmd.MarkFlagRequired("to-address")
+	_ = metastoreCopyAllCmd.Flags().String("schema-filter", "*", "filter for schemas to copy in regex")
+	_ = metastoreCopyAllCmd.Flags().String("table-filter", "*", "filter for tables to copy in regex")
```
Comment (nopcoder) [inline cmd/lakectl/cmd/metastore.go:258]: If filter syntax is regexp - '*' as default value is not a good one.
We can go with an empty value which means no filter or '.*'.

File: pkg/metastore/hive/ms_client.go [hunk 2]
```diff
@@ -41,6 +46,36 @@ func NewMSClient(ctx context.Context, addr string, secure bool) (*MSClient, erro
 	return msClient, nil
 }
 
+func CopyOrMergeAll(ctx context.Context, fromClient, toClient MSClient, schemaFilter, tableFilter, toBranch string) error {
```
Comment (nopcoder) [inline pkg/metastore/hive/ms_client.go:49]: `fromClient` and `toClient` structures: consider passing them by reference and not a copy.

File: pkg/metastore/hive/ms_client.go [hunk 3]
```diff
@@ -76,22 +111,27 @@ func (c *MSClient) Close() error {
 }
 
 func (c *MSClient) CopyOrMerge(fromDB, fromTable, toDB, toTable, toBranch, serde string, partition []string) error {
+	return c.CopyOrMergeToClient(fromDB, fromTable, toDB, toTable, toBranch, serde, partition, *c)
+}
+
+func (c *MSClient) CopyOrMergeToClient(fromDB, fromTable, toDB, toTable, toBranch, serde string, partition []string, toClient MSClient) error {
```
Comment (nopcoder) [inline pkg/metastore/hive/ms_client.go:117]: Consider calling the method - CopyOrMergeTo.
Without the Client, it is enough to understand that the operation accepts the target/destination.

File: pkg/metastore/hive/ms_client.go [hunk 4]
```diff
@@ -22,6 +22,11 @@ type ThriftHiveMetastoreClient interface {
 	AlterPartition(ctx context.Context, dbName string, tableName string, values *hive_metastore.Partition) (err error)
 	AddPartition(ctx context.Context, newPartition *hive_metastore.Partition) (r *hive_metastore.Partition, err error)
 	DropPartition(ctx context.Context, dbName string, tableName string, values []string, deleteData bool) (r bool, err error)
+	GetDatabase(ctx context.Context, name string) (r *hive_metastore.Database, err error)
+	GetDatabases(ctx context.Context, pattern string) (r []string, err error)
```
Comment (nopcoder) [inline pkg/metastore/hive/ms_client.go:26]: ListDatabases

File: pkg/metastore/hive/ms_client.go [hunk 5]
```diff
@@ -22,6 +22,11 @@ type ThriftHiveMetastoreClient interface {
...
+	GetAllDatabases(ctx context.Context) (r []string, err error)
+	CreateDatabase(ctx context.Context, database *hive_metastore.Database) (err error)
+	GetTables(ctx context.Context, dbName string, pattern string) (r []string, err error)
```
Comment (nopcoder) [inline pkg/metastore/hive/ms_client.go:29]: ListTables

File: cmd/lakectl/cmd/metastore.go (hunk 1 above)
Comment (guy-har) [reply cmd/lakectl/cmd/metastore.go:258]: It is filtered using hive metastore logic, which is somewhere between wildcard and regex. Passing an empty filter will return no results.

File: pkg/metastore/hive/ms_client.go (hunk 2 above)
Comment (guy-har) [reply pkg/metastore/hive/ms_client.go:49]: 👍

File: pkg/metastore/hive/ms_client.go (hunk 3 above)
Comment (guy-har) [reply pkg/metastore/hive/ms_client.go:117]: 👍

File: cmd/lakectl/cmd/metastore.go [hunk 6]
```diff
@@ -77,6 +77,46 @@ var metastoreCopyCmd = &cobra.Command{
 	},
 }
 
+var metastoreCopyAllCmd = &cobra.Command{
+	Use:   "copy-all",
+	Short: "copy from one metastore to another ",
```
Comment (nopcoder) [inline cmd/lakectl/cmd/metastore.go:82]: extra space at the end

File: pkg/metastore/hive/ms_client.go [hunk 7]
```diff
@@ -41,6 +46,36 @@ func NewMSClient(ctx context.Context, addr string, secure bool) (*MSClient, erro
...
+		database, err := fromClient.client.GetDatabase(ctx, dbName)
+		if err != nil {
+			return err
+		}
+		err = toClient.client.CreateDatabase(ctx, database)
+		if err != nil {
+			if _, ok := err.(*hive_metastore.AlreadyExistsException); !ok {
+				return err
+			}
+		}
```
Comment (nopcoder) [inline pkg/metastore/hive/ms_client.go:64]:
```suggestion
		var alreadyExistsErr *hive_metastore.AlreadyExistsException
		if errors.As(err, &alreadyExistsErr) {
			return err
		}
```

File: pkg/metastore/hive/ms_client.go [hunk 8]
```diff
@@ -41,6 +46,36 @@ func NewMSClient(ctx context.Context, addr string, secure bool) (*MSClient, erro
...
+		tablesNames, err := fromClient.client.GetTables(ctx, dbName, tableFilter)
+		if err != nil {
+			return err
+		}
+		for _, tableName := range tablesNames {
+			fmt.Printf("copy table  %s.%s\n", dbName, tableName)
```
Comment (nopcoder) [inline pkg/metastore/hive/ms_client.go:70]: remove the double space

**Review by nopcoder — APPROVED** [review]

**Pull request merged**